analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Binary.cs

```diff
@@ -758,4 +758,152 @@ public void Binary_Calculations_UpdatesNumberConstraint(string expression, int e
+    [DataRow("i >=  3 && j >=  5", 3 * 5, null)]
+    [DataRow("i >=  3 && j <= -5", null, 3 * -5)]
+    [DataRow("i <= -3 && j >=  5", null, -3 * 5)]
+    [DataRow("i <= -3 && j <= -5", -3 * -5, null)]
+    [DataRow("i ==  3 && j >=  5", 3 * 5, null)]
+    [DataRow("i ==  3 && j >= -5", 3 * -5, null)]
+    [DataRow("i ==  3 && j <=  5", null, 3 * 5)]
+    [DataRow("i ==  3 && j <= -5", null, 3 * -5)]
+    [DataRow("i == -3 && j >=  5", null, -3 * 5)]
+    [DataRow("i == -3 && j >= -5", null, -3 * -5)]
+    [DataRow("i == -3 && j <=  5", -3 * 5, null)]
+    [DataRow("i == -3 && j <= -5", -3 * -5, null)]
+    public void Binary_MultiplicationWithRange_UpdatesNumberConstraint(string expression, int? expectedMin, int?expectedMax)
+    {
+        var code = $$"""
+            if ({{expression}})
+            {
+                var value = i * j;
+                Tag("Value", value);
+            }
+            """;
+        var validator = SETestContext.CreateCS(code, "int? i, int? j").Validator;
+        validator.ValidateTag("Value", x => x.Should().HaveOnlyConstraints(NumberConstraint.From(expectedMin, expectedMax)));
+    }
+
+    [DataTestMethod]
+    [DataRow("i >=  3 && j >= -5")]
+    [DataRow("i >=  3 && j <=  5")]
+    [DataRow("i >= -3 && j >=  5")]
+    [DataRow("i >= -3 && j >= -5")]
+    [DataRow("i >= -3 && j <=  5")]
+    [DataRow("i >= -3 && j <= -5")]
+    [DataRow("i <=  3 && j >=  5")]
+    [DataRow("i <=  3 && j >= -5")]
+    [DataRow("i <=  3 && j <=  5")]
+    [DataRow("i <=  3 && j <= -5")]
+    [DataRow("i <= -3 && j >= -5")]
+    [DataRow("i <= -3 && j <=  5")]
+    public void Binary_MultiplicationWithRange_NoNumberConstraint(string expression)
+    {
+        var code = $$"""
+            if ({{expression}})
+            {
+                var value = i * j;
+                Tag("Value", value);
+            }
+            """;
+        var validator = SETestContext.CreateCS(code, "int? i, int? j").Validator;
+        validator.ValidateTag("Value", x => x.Should().BeNull());
+    }
+
+    [DataTestMethod]
+    [DataRow(0b0101, 0b0101, 0b0101)]
+    [DataRow(0b0101, 0b0001, 0b0001)]
+    [DataRow(0b1010, 0b0110, 0b0010)]
+    [DataRow(0b1010, 0b0000, 0b0000)]
+    [DataRow(5, -5, 1)]
+    [DataRow(5, -4, 4)]
+    [DataRow(-5, -5, -5)]
+    [DataRow(-5, -4, -8)]
+    public void Binary_BitAndWithFixedValues_UpdatesNumberConstraint(int left, int right, int expected)
+    {
+        var code = $"""
+            var left = {left};
+            var right = {right};
+            var value = left & right;
+            Tag("Value", value);
+            """;
+        var validator = SETestContext.CreateCS(code).Validator;
+        validator.ValidateTag("Value", x => x.Should().HaveOnlyConstraints(ObjectConstraint.NotNull, NumberConstraint.From(expected)));
+    }
+
+    [DataTestMethod]
+    [DataRow("i >=  3 && j >=  5", 0, null)]
+    [DataRow("i >=  3 && j >= -5", 0, null)]
+    [DataRow("i >=  3 && j <=  5", 0, null)]
```

Comment:
For two positive numbers, if there's a `Max`, `and` is `min(left.Max, right.Max)`. Because if I take the bigger number, the smaller can only kill some bits.
```suggestion
    [DataRow("i >=  3 && j <=  5", 0, 5)]
```

Reply:
If `j` is `-1` the result will be the same as `i` => the max is unlimited.